### PR TITLE
Select operators.

### PR DESCRIPTION
**Pull request by sunfishcode**
This adds select (analogous to conditional move) operators on all the local types.

### CODE REVIEW

**Comment by sunfishcode** [comment]
It was pointed out that `select` can be expressed as a type-parametric operator. Updated.

**Comment by rossberg-chromium** [comment]
Isn't this simply (a strict version of) `if`?

**Comment by titzer** [comment]
@rossberg-chromium yes, indeed. It can be implemented as a conditional move instruction.

**Comment by sunfishcode** [comment]
Updated to describe the operand order and semantics of `select` in more detail.

**Comment by titzer** [comment]
lgtm

**Comment by rossberg-chromium** [comment]
@titzer, sure, but i assume you'd use conditional move instructions for suitable `if`s as well? So what does a separate operator add?

**Comment by sunfishcode** [comment]
@rossberg-chromium When a consumer sees an `if` operator, it can immediately start setting up the bookkeeping needed to create basic blocks or an SSA scope or whatever it needs to prepare to decode the rest of the body. `select` can be translated directly to a conditional move, which is a very common hardware instruction.

**Comment by qwertie** [comment]
+1 (btw, unto whom is destowed the power of the lgtm?)

**Comment by rossberg-chromium** [comment]
@sunfishcode, what you're saying sounds like it's wanted only for the sake of micro-optimising some decoders. Is that really the primary motivation?

**Comment by qwertie** [comment]
I see this as shifting optimization work (the decision between branch and conditional moves) from browser-time to compile-time. As long as compilers tend to make the same decision for different ISAs, it makes sense.

**Comment by sunfishcode** [comment]
It's also convenient for producers which already have a strict-evaluation select operation [eg.](http://llvm.org/docs/LangRef.html#i-select) [eg.](http://pp.ipd.kit.edu/firm/Nodes_latest#Mux). It makes it easier to work across multiple levels of the system (producer, wasm, consumer, machine code) if we can minimize the number of intermediate semantic translations/optimizations needed in common cases.

**Comment by titzer** [comment]
The use of a select operator is a clear signal from a producer to a consumer that a conditional move is a best implementation strategy for the given condition. It doesn't compel the consumer to do so, and a simple and correct implementation is to just emit an normal high-level if after evaluating all operands. 

Implementation burden is not high, maps to a single instruction on multiple machines, expresses clear optimization hint, so, I think it makes sense to add this.

File: AstSemantics.md [hunk 1]
```diff
@@ -534,6 +534,13 @@ Truncation from floating point to integer where IEEE 754-2008 would specify an
 invalid operation exception (e.g. when the floating point value is NaN or
 outside the range which rounds to an integer in range) traps.
 
+## Type-parameterized operators.
+
+  * `select`: a ternary operator with a boolean (i32) condition and two
+    additional operands with the parameterized type. `select` returns the
+    the first type-parameterized operand if the condition operand is non-zero,
+    or the second type-parameterized operand otherwise.
```
Comment (lukewagner) [inline AstSemantics.md]: I think it'd be simpler to not say "parameterized" at all and just have a sentence at the end saying the parameters must have the same value type.

**Comment by lukewagner** [comment]
Generally lgtm

**Comment by sunfishcode** [comment]
Updated to avoid saying "type-parameterized operands", and to just say that the two operands to select between must have the same type as each other.

**Comment by sunfishcode** [comment]
Merging, with lgtm's and +1's above.

**Pull request merged**